### DISCUSSION TITLE
How to use Razzle with BrowsersList?

### DISCUSSION THREAD

**Discussion by silviubogan** [jaredpalmer/razzle#1698] [Q&A] [answered]
Hello!

In my Razzle app (TS + React) I tried to put different browserslist configuration in package.json but the bundle size did not change significantly (less than 30 B of changes in the bundle). Is it sure that the configuration from package.json is all I need to change how WebPack produces the bundle?

Thanks.

**Comment by fivethreeo** [comment]
Browserslist only affects postcss,  but should be enough to change package.json

**Comment by fivethreeo** [comment]
Should  affect babel too but does not now i think, babel loader needs fixing in razzle it seems

**Comment by fivethreeo** [comment]
https://github.com/jaredpalmer/razzle/blob/master/packages/razzle/config/loadRazzleConfig.js#L44
https://github.com/jaredpalmer/razzle/blob/master/packages/razzle/config/createConfigAsync.js#L291
https://github.com/jaredpalmer/razzle/blob/a10a9948941897752457da8c04c6022191dd09f1/packages/razzle/config/babel-loader/razzle-babel-loader.js#L119
https://github.com/jaredpalmer/razzle/blob/a10a9948941897752457da8c04c6022191dd09f1/packages/babel-preset-razzle/index.js#L31

So you can pass it down to the babel preset by modifying babelrule

**Reply by silviubogan** [reply]
Sorry for the dumb question, but what is babelrule and how can I change it? Should make a .babelrc file and put the browserslist inside it?

**Comment by fivethreeo** [comment]
https://github.com/jaredpalmer/razzle/blob/a10a9948941897752457da8c04c6022191dd09f1/packages/razzle/config/createConfigAsync.js#L286

**Comment by fivethreeo** [comment]
use modifyWebpackOptions

razzleOptions.browserslist
webpackOptions.babelRule.options['preset-env']


https://razzlejs.org/docs/customization#extending-webpack
https://razzlejs.org/docs/customization#writing-plugins

**Comment by fivethreeo** [comment]
Hm, my idea won't work ... :(

https://babeljs.io/docs/en/babel-preset-env#browserslist-integration

**Comment by fivethreeo** [comment]
Or, maybe ..

webpackOptions.babelRule.options['preset-env'].targets = razzleOptions.browserslist

**Reply by silviubogan** [reply]
In `modifyWebpackConfig`, `opts.options.webpackOptions.babelRule` is like this:

```js
{
  test: /\.(js|jsx|mjs|ts|tsx)$/,
  include: [
    '/path/to/my/project/dir/src',
    '/path/to/my/project/dir/src',
    /react-syntax-highlighter/,
    /react-virtualized/,
    /react-select-virtualized/,
    /tiny-slider/
  ],
  use: [
    {
      loader: '/path/to/my/project/dir/.yalc/razzle/config/babel-loader/razzle-babel-loader.js',
      options: [Object],
      ident: 'razzle-babel-loader'
    }
  ]
}
```

so the following code throws an Error when running `npm start`:

```js
modifyWebpackConfig(opts) {
  const config = opts.webpackConfig;
  const options = opts.options.webpackOptions;

  if (opts.env.target === "web") {
    options.babelRule.options['preset-env'].targets = opts.options.razzleOptions.browserslist;
  }
  // ...
```

because `options.babelRule.options` is undefined.

**Reply by silviubogan** [reply]
It seems that there can't be both `use` and `options` inside `babelRule`.

**Reply by silviubogan** [reply]
I am currently looking into `razzle-babel-loader.js` to find how can I send browserslist to Babel.

**Comment by fivethreeo** [comment]
But only if target is web

**Comment by fivethreeo** [comment]
modifyWebpackOptions not Config 😀

**Reply by silviubogan** [reply]
```js
modifyWebpackOptions({ env: { target, dev }, options: { pluginOptions, razzleOptions, webpackOptions } }) {
  webpackOptions.notNodeExternalResMatch = (request) => new RegExp('react-syntax-highlighter|'
    + 'react-virtualized|'
    + 'react-select-virtualized|'
    + 'tiny-slider').test(request);

  webpackOptions.babelRule.include = webpackOptions.babelRule.include.concat([
    /react-syntax-highlighter/,
    /react-virtualized/,
    /react-select-virtualized/,
    /tiny-slider/,
  ]);

  if (target === "web") {
    webpackOptions.babelRule.use[0].options.targets = razzleOptions.browserslist;
  }

  return webpackOptions;
},
```

> Error: Unknown option: `.targets`. Check out https://babeljs.io/docs/en/babel-core/#options for more information about options.

If I directly assign to `use[0].targets`:

> Invalid configuration object. Webpack has been initialized using a configuration object that does not match the API schema.
> - configuration.module.rules[0].use[0] has an unknown property 'targets'. These properties are valid:
>   object { ident?, loader?, options? }

**Comment by fivethreeo** [comment]
And babelRule.use[0].options

**Reply by silviubogan** [reply]
This does not change the dev bundle size at all:

```js
webpackOptions.babelRule.use.options = { targets: razzleOptions.browserslist };
```

But at least it does not throw exceptions.

**Comment by fivethreeo** [comment]
I looked closer at it, have been giving you bad advice, has to be done in .babelrc ;)

**Comment by fivethreeo** [comment]
https://razzlejs.org/docs/customization#customizing-babel-config

**Reply by silviubogan** [reply]
I've put these in .babelrc:

```json
{
  "presets": [
    "razzle/babel"
  ],
  "plugins": [],
  "targets": "> 0.25%, not dead"
}
```

and it tells me that `targets` is invalid property

**Comment by fivethreeo** [comment, marked as answer]
```json

{
  "presets": [
    [
      "razzle/babel",
      {
        "preset-env": {"targets": "> 0.25%, not dead"},
      }
    ]
  ]
}
```
